ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java

```diff
@@ -1521,12 +1570,22 @@ private static int estimateNumBuckets(JoinOperator joinOp, boolean useOpTraits)
     int estimatedBuckets = -1;
 
     for (Operator<? extends OperatorDesc>parentOp : joinOp.getParentOperators()) {
-      if (parentOp.getOpTraits().getNumBuckets() > 0) {
-        numBuckets = (numBuckets < parentOp.getOpTraits().getNumBuckets()) ?
-            parentOp.getOpTraits().getNumBuckets() : numBuckets;
+      final OpTraits parentOpTraits = parentOp.getOpTraits();
+      if (parentOpTraits.getNumBuckets() > 0) {
+        numBuckets = (numBuckets < parentOpTraits.getNumBuckets()) ?
+            parentOpTraits.getNumBuckets() : numBuckets;
+      }
+
+      if (!(parentOp instanceof ReduceSinkOperator)) {
```

Comment:
why not filter upfront

Reply:
This tried to keep the original logic as much as possible, but I think your question should be valid.
I tried to move it and am monitoring CI
https://github.com/apache/hive/pull/5409/commits/c6aa9d2d0a81ff2e8834e1350a18405e593fe212